### PR TITLE
Fix cleanClosure() on recursive function calls.

### PR DESCRIPTION
**Pull request by hlin09**
1. Grouped several test cases together to condense the test code. Added test for def-use chain with interleaving function calls, and recursive function calls. 
2. Fixed the issue with recursive function calls, which led to infinite recursion.
3. Replace global `defVars` (holding the local var names) with local one to `cleanClosure()`. The previous version will clear the `defVars` once `processClosure()` calls cleanClosure() on processing functions. (https://github.com/amplab-extras/SparkR-pkg/blob/master/pkg/R/utils.R#L312)

### CODE REVIEW

File: pkg/R/utils.R [hunk 1]
```diff
@@ -305,11 +305,27 @@ processClosure <- function(node, oldEnv, argNames, newEnv) {
               !(nodeChar %in% getNamespaceExports("SparkR")))) {  # Only include SparkR internals.
           # Set parameter 'inherits' to FALSE since we do not need to search in
           # attached package environments.
-          if (exists(nodeChar, envir = func.env, inherits = FALSE)) {
+          if (tryCatch(exists(nodeChar, envir = func.env, inherits = FALSE),
```
Comment (shivaram) [inline pkg/R/utils.R]: what kinds of errors can `exists` throw ? I am just wondering why we need a `tryCatch` here

File: pkg/R/utils.R [hunk 2]
```diff
@@ -305,11 +305,27 @@ processClosure <- function(node, oldEnv, argNames, newEnv) {
...
+                       error = function(e) { FALSE })) {
             obj <- get(nodeChar, envir = func.env, inherits = FALSE)
-            if (is.function(obj)) {  
-              # if the node is a function call, recursively clean its closure.
-              obj <- cleanClosure(obj)
+            if (is.function(obj)) {  # If the node is a function call.
+              funcList <- mget(nodeChar, envir = checkedFuncs, inherits = F, 
+                               ifnotfound = list(list(NULL)))[[1]]
```
Comment (shivaram) [inline pkg/R/utils.R]: Hmm - just so I understand correctly if the funcList is `list(list(NULL))` we will go inside the `sapply` but the `identical(func, obj)` will be `FALSE` ?

**Comment by shivaram** [comment]
Thanks @hlin09 - This is a very useful change. I just had a couple of minor questions about the code inline.

File: pkg/R/utils.R (hunk 2 above)
Comment (hlin09) [inline pkg/R/utils.R]: Yes, you are right. I admit that it is less readable, but I was trying to void extra check for errors returned by `get` if there is no such `nodeChar`. So I use `mget` with default value `list(list(NULL))`.

File: pkg/R/utils.R (hunk 1 above)
Comment (hlin09) [inline pkg/R/utils.R]: At least one error I came across was if the `nodeChar` is `""`, the `get` return a error of `invalid argument`. So I thought that should be a `FALSE` in any case of error.

Comment (shivaram) [inline pkg/R/utils.R]: Ah yes - I see that passing in a empty string does throw an error. Having a `tryCatch` is fine here.

**Comment by shivaram** [comment]
Thanks @hlin09 - LGTM. I'll merge this into master and cherry-pick it into `sparkr-sql`.

**Pull request merged**